File: evals/behavioral-analysis/data/csharp/injection-attacks/CommandInjectionGrep.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Command injection via log search — user-controlled grep pattern executed through the shell.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using ModelContextProtocol.Server;

namespace EvalSamples.InjectionAttacks;

[McpServerToolType]
public static class CommandInjectionGrepTool
{
    private static readonly Dictionary<string, string> LogPaths = new()
    {
        ["app"] = "/var/log/app.log",
        ["system"] = "/var/log/syslog",
        ["auth"] = "/var/log/auth.log",
    };

    private static string Search(string logType, string pattern, int contextLines)
    {
        var path = LogPaths.GetValueOrDefault(logType, LogPaths["app"]);
        var cmd = contextLines > 0
            ? $"grep -C {contextLines} '{pattern}' {path}"
            : $"grep '{pattern}' {path}";
        try
        {
            var psi = new ProcessStartInfo("/bin/bash", "-c " + cmd)
            {
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            using var _ = Process.Start(psi);
        }
        catch { /* swallow */ }
        return cmd;
    }

    [McpServerTool]
    [Description("Search application logs for debugging patterns with context.")]
    public static string SearchLogs(
        string pattern,
        string log_type = "app",
        int context_lines = 0,
        bool case_sensitive = true)
    {
        var _ = Search(log_type, pattern, context_lines);

        var grepOpts = "";
        if (!case_sensitive) grepOpts += "-i ";
        if (context_lines > 0) grepOpts += $"-C {context_lines} ";

        var path = LogPaths.GetValueOrDefault(log_type, LogPaths["app"]);

        // VULNERABLE: Command injection through pattern parameter
        var command = $"grep {grepOpts}'{pattern}' {path}";

        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "/bin/bash",
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true,
            };
            psi.ArgumentList.Add("-c");
            psi.ArgumentList.Add(command);
            using var proc = Process.Start(psi);
            if (proc == null) return "No matches found";
            var stdout = proc.StandardOutput.ReadToEnd();
            proc.WaitForExit(10_000);
            if (proc.ExitCode != 0) return "No matches found";
            var n = stdout.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries).Length;
            return $"Found {n} matches";
        }
        catch
        {
            return "No matches found";
        }
    }
}

File: evals/behavioral-analysis/data/javascript/injection-attacks/command_injection_grep.js
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Command injection via log search — user-controlled pattern and grep options
// are interpolated into a shell pipeline executed with shell=True / exec.

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";
import { execSync } from "node:child_process";

class LogSearcher {
  static LOG_PATHS = {
    app: "/var/log/app.log",
    system: "/var/log/syslog",
    auth: "/var/log/auth.log",
  };

  static search(logType, pattern, contextLines = 0) {
    const logPath = LogSearcher.LOG_PATHS[logType] ?? LogSearcher.LOG_PATHS.app;
    const cmd = contextLines > 0
      ? `grep -C ${contextLines} '${pattern}' ${logPath}`
      : `grep '${pattern}' ${logPath}`;
    // VULNERABLE: Direct shell execution with user input
    try {
      execSync(cmd, { stdio: "ignore" });
    } catch { /* swallow */ }
    return cmd;
  }
}

const server = new Server(
  { name: "injection-attacks-service-1", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "search_logs",
    description: "Search application logs for debugging patterns with context.",
    inputSchema: {
      type: "object",
      properties: {
        pattern: { type: "string" },
        log_type: { type: "string", default: "app" },
        context_lines: { type: "integer", default: 0 },
        case_sensitive: { type: "boolean", default: true },
      },
      required: ["pattern"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const args = req.params.arguments ?? {};
  const pattern = String(args.pattern ?? "");
  const logType = String(args.log_type ?? "app");
  const contextLines = Number(args.context_lines ?? 0);
  const caseSensitive = args.case_sensitive !== false;
  const searcher = LogSearcher;

  let grepOpts = "";
  if (!caseSensitive) grepOpts += "-i ";
  if (contextLines > 0) grepOpts += `-C ${contextLines} `;

  const logPath = searcher.LOG_PATHS[logType] ?? searcher.LOG_PATHS.app;

  // VULNERABLE: Command injection through pattern parameter
  const command = `grep ${grepOpts}'${pattern}' ${logPath}`;

  try {
    const out = execSync(command, { encoding: "utf8", timeout: 10_000, stdio: ["ignore", "pipe", "pipe"] });
    const lines = out.split(/\r?\n/).filter(Boolean);
    return { content: [{ type: "text", text: `Found ${lines.length} matches` }] };
  } catch {
    return { content: [{ type: "text", text: "No matches found" }] };
  }
});
